File: Lokesh_Saana/986-interval-list-intersections.py
# https://leetcode.com/problems/interval-list-intersections/

class Solution:
    def intervalIntersection(self, firstList: List[List[int]], secondList: List[List[int]]) -> List[List[int]]:
        i = j = 0
        result = []
        while i < len(firstList) and j < len(secondList):
            a_left, a_right = firstList[i]
            b_left, b_right = secondList[j]
            if a_left <= b_right and b_left <= a_right:
                result.append([max(a_left, b_left), min(a_right, b_right)])
            if a_right <= b_right:
                i += 1
            else:
                j += 1
        return result

File: Ayush-Gupta/986. Interval List Intersections.cpp
class Solution {
public:
    vector<vector<int>> intervalIntersection(vector<vector<int>>& a, vector<vector<int>>& b) {
        vector<vector<int>> ans;
        int i = 0, j = 0, n = a.size(), m = b.size(), lo, hi;
        while(i < n && j < m) {
            lo = max(a[i][0], b[j][0]);
            hi = min(a[i][1], b[j][1]);
            if(lo <= hi) ans.push_back({lo, hi});
            if(a[i][1] < b[j][1]) i++;
            else j++;
        }
        return ans;
    }
};